File: frontend/services/doctors.ts
import { api } from "@/lib/api";
import type { NearbyPlace, PlaceType } from "@/types";

export async function findNearbyPlaces(
  latitude: number,
  longitude: number,
  placeType?: PlaceType,
  radiusMeters = 50000,
  search?: string
): Promise<NearbyPlace[]> {
  const { data } = await api.post("/api/v1/doctors/nearby", {
    latitude,
    longitude,
    radius_meters: radiusMeters,
    place_type: placeType,
    search: search?.trim() || undefined,
  });

  return data.places.map((p: any) => ({
    id: p.id,
    name: p.name,
    type: p.type,
    latitude: p.latitude,
    longitude: p.longitude,
    distanceKm: p.distance_km,
    address: p.address,
    phone: p.phone,
  }));
}

File: backend/app/api/routes/doctors.py
from fastapi import APIRouter, Depends, HTTPException, Request

from app.core.rate_limit import limiter, DEFAULT_RATE
from app.core.security import get_current_user, AuthenticatedUser
from app.models.schemas import NearbyPlacesRequest, NearbyPlacesResponse
from app.services.overpass_service import find_nearby_places

router = APIRouter(prefix="/api/doctors", tags=["doctors"])


@router.post("/nearby", response_model=NearbyPlacesResponse)
@limiter.limit(DEFAULT_RATE)
async def nearby(
    request: Request,
    payload: NearbyPlacesRequest,
    user: AuthenticatedUser = Depends(get_current_user),
):
    try:
        places = await find_nearby_places(
            payload.latitude,
            payload.longitude,
            payload.radius_meters,
            payload.place_type,
            payload.search,
        )
    except Exception as exc:  # Overpass occasionally times out / rate-limits
        raise HTTPException(
            status_code=502, detail="Could not reach the map data provider. Please try again shortly."
        ) from exc

    return NearbyPlacesResponse(places=places[:40])
